airflow/api_fastapi/execution_api/routes/xcoms.py

```diff
@@ -175,7 +160,7 @@ def set_xcom(
     task_id: str,
     key: str,
     value: Annotated[
-        str,
+        JsonValue,
```

Comment:
fyi: We would still do multiple serialization/ de-serialization steps that we should consolidate -- even after this PR.

**Storing Data** (for GO-client example):

1. Go serializes the object (json.Marshal) → Sends it as an HTTP request.
2. FastAPI deserializes it (JsonValue) → Converts it into a Python dict.
3. SQLAlchemy stores it (Column(JSON)) → Serializes it back into JSON in the database.

https://github.com/apache/airflow/blob/8f63b828ace09b1095229f22b0c6d1f0f85ea81a/airflow/models/xcom.py#L83





**Retrieving Data**:

1. SQLAlchemy loads it as a Python dict from the DB.
2. FastAPI serializes it back into JSON.
3. Go receives it and deserializes it (`json.Unmarshal` or a different library).

Reply:
Yeah, we could look at improving this in the future but it's not critical for Day 0.

Reply:
That's true, we can revisit it later again if we need to.

By multi serialisation i meant that we stored a multi serialised value in the DB. This also lead to a difference from airflow 2 format, which means those xcoms cant be retrieved.

The current logic does multiple (times) serialisation, but stores singly serialised value in DB.

Reply:
Resolving for now, if anything, lets discuss in a follow up